File: Week_08-位运算-布隆过滤-排序/146.lru缓存机制.cpp
/*
 * @lc app=leetcode.cn id=146 lang=cpp
 *
 * [146] LRU缓存机制
 */

/*
LRU缓存操作原理：
头部放最新访问数据，尾部放很久前访问的数据；
1.访问数据，查询待访问数据地址，若有，返回此地址中存储的值，并把它放入头部；
2.存入数据，查询待访问数据地址，若有，把它调整到头部；若无，把它插入到头部，并判断cache是否溢出（超过其capacity），若溢出，删掉尾部；

LRU的get，put操作的两个关键元素：key 和 value， 理解成cache的地址索引和其中存储的值；
note，为区别哈希列表和cache，在下述哈希列表中，用键和值表述hash_map的key和value；
双端链表，cache的数据存储。 key，对应于cache的key，就是cache的地址索引；value，对应于cache的value，实际存放的数据；
哈希列表，键，对应于cache的key，就是cache的地址索引； 值，对应于链表的结点的地址；

为什么不用单链表？
单链表，只定义了head，如果想要快速访问tail，要逐个从头部到尾部遍历，时间复杂度O(n)；不适用于用O(1)访问和删除尾部的场景；
而双链表，定义了head和tail，访问和删除尾部时，就比较迅速了;
*/
// @lc code=start
#include<unordered_map>
using namespace std;
//双端链表，具体的cache实现结构
struct DlinkNode {     
    int key, value;
    DlinkNode* prev;
    DlinkNode* next;
    DlinkNode() : key(0), value(0), prev(nullptr), next(nullptr) {}   //构造，无参数，初始化
    DlinkNode(int _key, int _value): key(_key), value(_value), prev(nullptr), next(nullptr) {} //构造，有参数，初始化
};

class LRUCache {
private:
    unordered_map<int, DlinkNode*> cache;   //哈希map 存放索引cache的信息
    DlinkNode* head;
    DlinkNode* tail;
    int size;
    int capacity;
public:
    LRUCache(int _capacity): capacity(_capacity), size(0) {
        head = new DlinkNode();
        tail = new DlinkNode();
        head->next = tail;
        tail->prev = head;
    }
    
    int get(int key) {
        if (!cache.count(key)) {
            return -1;
        }
        DlinkNode* node = cache[key];
        moveToHead(node);
        return node->value;   //获取cache数据，链表中的值；

    }
    
    void put(int key, int value) {
        if (!cache.count(key)) {
            DlinkNode* node = new DlinkNode(key, value);
            cache[key] = node;
            addToHead(node);
            ++size;
            if (size > capacity) {
                DlinkNode* removed = removeTail();
                cache.erase(removed->key);
                delete removed;
                --size;
            }
        }
        else {
            DlinkNode* node = cache[key];
            node->value = value;
            moveToHead(node);
        }
    }
    void addToHead(DlinkNode* node) {
        node->prev = head;
        node->next = head->next;
        head->next->prev = node;
        head->next = node;
    }
    void removeNode(DlinkNode* node) {
        node->prev->next = node->next;
        node->next->prev = node->prev;
    }
    void  moveToHead(DlinkNode* node) {
        removeNode(node);
        addToHead(node);
    }
    DlinkNode* removeTail() {
        DlinkNode* node = tail->prev;
        removeNode(node);
        return node;
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */
// @lc code=end

File: Week_08-位运算-布隆过滤-排序/146.lru-缓存机制.py
#
# @lc app=leetcode.cn id=146 lang=python
#
# [146] LRU 缓存机制
#

# LRU 缓存机制可以通过哈希表辅以双向链表实现，我们用一个哈希表和一个双向链表维护所有在缓存中的键值对。
# 双向链表按照被使用的顺序存储了这些键值对，靠近头部的键值对是最近使用的，而靠近尾部的键值对是最久未使用的。
# 哈希表即为普通的哈希映射（HashMap），通过缓存数据的键映射到其在双向链表中的位置。
# 这样以来，我们首先使用哈希表进行定位，找出缓存项在双向链表中的位置，随后将其移动到双向链表的头部，即可在 
# O(1) 的时间内完成 get 或者 put 操作。具体的方法如下：
# 对于 get 操作，首先判断 key 是否存在：
#   如果 key 不存在，则返回 −1；
#   如果 key 存在，则 key 对应的节点是最近被使用的节点。通过哈希表定位到该节点在双向链表中的位置，并将其移动到双向链表的头部，最后返回该节点的值。
# 对于 put 操作，首先判断 key 是否存在：
#   如果 key 不存在，使用 key 和 value 创建一个新的节点，在双向链表的头部添加该节点，并将 key 和该节点添加进哈希表中。然后判断双向链表的节点数是否超出容量，如果超出容量，则删除双向链表的尾部节点，并删除哈希表中对应的项；
#   如果 key 存在，则与 get 操作类似，先通过哈希表定位，再将对应的节点的值更新为 value，并将该节点移到双向链表的头部。
# 上述各项操作中，访问哈希表的时间复杂度为 O(1)，在双向链表的头部添加节点、在双向链表的尾部删除节点的复杂度也为 O(1)。
# 而将一个节点移到双向链表的头部，可以分成「删除该节点」和「在双向链表的头部添加节点」两步操作，都可以在 O(1) 时间内完成。
# 小技巧：在双向链表的实现中，使用一个伪头部（dummy head）和伪尾部（dummy tail）标记界限，这样在添加节点和删除节点的时候就不需要检查相邻的节点是否存在。

# @lc code=start
class DLinkedNode:
    def __init__(self, key=0, value=0):
        self.key = key
        self.value  = value
        self.prev = None
        self.next = None

class LRUCache(object):

    def __init__(self, capacity):
        """
        :type capacity: int
        """
        self.cache = dict()
        # 使用伪头部和伪尾部节点
        self.head = DLinkedNode()
        self.tail = DLinkedNode()
        self.head.next = self.tail
        self.tail.prev = self.head
        self.capacity = capacity
        self.size = 0


    def get(self, key):
        """
        :type key: int
        :rtype: int
        """
        if key not in self.cache:
            return -1
        # 如果key存在，县通过哈希表定位它的node，再移到头部，并返回node的value
        node = self.cache[key]
        self.moveToHead(node)
        return node.value


    def put(self, key, value):
        """
        :type key: int
        :type value: int
        :rtype: None
        """
        if key not in self.cache:
            # 如果key 不存在，创建一个新的key:value节点
            node = DLinkedNode(key, value)
            # 新节点添加进哈希表。哈希表中存的是key: node
            self.cache[key] = node
            # 添加到双向链表头部
            self.addToHead(node)
            self.size += 1
            if self.size > self.capacity:
                # 如果超出容量，删除双向链表尾部节点
                removed = self.removeTail()
                # 相应地删除哈希表中对应的key项
                self.cache.pop(removed.key)
                self.size -= 1
        else:
            # 如果key存在，先通过哈希表定位它的node，再修改value，再移到头部
            node = self.cache[key]
            node.value = value
            self.moveToHead(node)
    
    # 双向链表基础操作--添加头
    def addToHead(self, node):
        node.prev = self.head
        node.next = self.head.next
        self.head.next.prev = node
        self.head.next = node
    # 双向链表基础操作--删除节点
    def removeNode(self, node):
        node.prev.next = node.next
        node.next.prev = node.prev
    # 双向链表基础操作--把节点移动到头
    def moveToHead(self, node):
        self.removeNode(node)
        self.addToHead(node)
    # 双向链表基础操作--删除尾节点
    def removeTail(self):
        node = self.tail.prev
        self.removeNode(node)
        return node

# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)
# @lc code=end

